### ISSUE TITLE
Compilation failed with GCC 7

### ISSUE THREAD

**Issue by zorun** [msgpack/msgpack-c#600]
See the error below when compiling 2.1.2 with GCC 7, related to `-Werror=maybe-uninitialized`.  Maybe this option should be turned off for tests?  It seems like it's not the first time it causes compilation issues (#494).

```
Scanning dependencies of target msgpack_container
[ 53%] Building CXX object test/CMakeFiles/msgpack_container.dir/msgpack_container.cpp.o
/build/msgpack-c/src/msgpack-2.1.2/test/cases.cpp: In member function ‘virtual void cases_format_Test::TestBody()’:
/build/msgpack-c/src/msgpack-2.1.2/test/cases.cpp:31:32: error: ‘oh_compact.msgpack::v2::object::<anonymous>.msgpack::v1::object::via.msgpack::v1::object::union_type::str.msgpack::v1::object_str::ptr’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         msgpack::object_handle oh_compact;
                                ^~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [test/CMakeFiles/cases.dir/build.make:63: test/CMakeFiles/cases.dir/cases.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1385: test/CMakeFiles/cases.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by redboltz** [comment]
@zorun , thank you for reporting the issue.
Unfortunately, I couldn't reproduce the error, so far.

```
Scanning dependencies of target msgpack_container
[ 53%] Building CXX object test/CMakeFiles/msgpack_container.dir/msgpack_container.cpp.o
/build/msgpack-c/src/msgpack-2.1.2/test/cases.cpp: In member function ‘virtual void cases_format_Test::TestBody()’:
/build/msgpack-c/src/msgpack-2.1.2/test/cases.cpp:31:32: error: ‘oh_compact.msgpack::v2::object::<anonymous>.msgpack::v1::object::via.msgpack::v1::object::union_type::str.msgpack::v1::object_str::ptr’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         msgpack::object_handle oh_compact;
                                ^~~~~~~~~~
cc1plus: all warnings being treated as errors
```

It seems that the error is occurred on cases.cpp:31. But your log is building msgpack_container.cpp. Are you building parallel? 

Could you tell me parameters of cmake and make?

**Comment by zorun** [comment]
Hi,

Indeed, it's building in parallel.  I can reproduce more simply:

```
# Install gtest
cmake -DCMAKE_INSTALL_PREFIX=/usr -DMSGPACK_CXX11=ON -DMSGPACK_BUILD_EXAMPLES=OFF .
make -j1
```

Result:

```
Scanning dependencies of target cases
[ 31%] Building CXX object test/CMakeFiles/cases.dir/cases.cpp.o
/tmp/msgpack-2.1.2/test/cases.cpp: In member function ‘virtual void cases_format_Test::TestBody()’:
/tmp/msgpack-2.1.2/test/cases.cpp:31:32: error: ‘oh_compact.msgpack::v2::object::<anonymous>.msgpack::v1::object::via.msgpack::v1::object::union_type::str.msgpack::v1::object_str::ptr’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         msgpack::object_handle oh_compact;
                                ^~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [test/CMakeFiles/cases.dir/build.make:63: test/CMakeFiles/cases.dir/cases.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1482: test/CMakeFiles/cases.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

Versions:

* gcc (GCC) 7.1.1 20170528
* gtest 1.8.0

**Comment by redboltz** [comment]
I'm using the same version of gcc.

```
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/7.1.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc-multilib/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release
Thread model: posix
gcc version 7.1.1 20170528 (GCC) 
```

And using gtest 1.7.0.
I experienced some linking problem on multilib platform with gtest 1.8.0 so I use 1.7.0.

Anyway, the warning doesn't seems to be harmful. So I added warning suppression code. Could you check #602?

**Comment by redboltz** [comment]
@zorun , is your problem fixed by #602?

**Comment by zorun** [comment]
Hi,

I ended up removing `-Werror` for the tests, because it seems to bring more problems that it solves:

```
--- test/CMakeLists.txt.old	2017-06-11 17:11:18.413443078 +0200
+++ test/CMakeLists.txt	2017-06-11 17:11:39.383317557 +0200
@@ -80,7 +80,7 @@
     )
     ADD_TEST (${source_file_we} ${source_file_we})
     IF ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
-        SET_PROPERTY (TARGET ${source_file_we} APPEND_STRING PROPERTY COMPILE_FLAGS "-Wall -Wextra -Werror -g -O3 ")
+        SET_PROPERTY (TARGET ${source_file_we} APPEND_STRING PROPERTY COMPILE_FLAGS "-Wall -Wextra -g -O3 ")
     ENDIF ()
     IF ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
         SET_PROPERTY (TARGET ${source_file_we} APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-mismatched-tags")
```

However, I just tested #602 , and it fixes this particular issue.  But then there is another compilation error afterwards:

```
[ 45%] Building CXX object test/CMakeFiles/msgpack_stream.dir/msgpack_stream.cpp.o
In file included from /usr/include/gtest/gtest.h:1874:0,
                 from /build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp:7:
/build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp: In member function ‘virtual void MSGPACKC_simple_buffer_v4raw_fix_l_Test::TestBody()’:
/build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp:978:24: error: argument 1 null where non-null expected [-Werror=nonnull]
     EXPECT_EQ(0, memcmp(str, obj.via.str.ptr, str_size));
                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /build/msgpack-c/src/msgpack-2.1.2/include/msgpack/pack_define.h:15:0,
                 from /build/msgpack-c/src/msgpack-2.1.2/include/msgpack/pack.h:13,
                 from /build/msgpack-c/src/msgpack-2.1.2/include/msgpack.h:19,
                 from /build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp:1:
/usr/include/string.h:66:12: note: in a call to function ‘int memcmp(const void*, const void*, size_t)’ declared here
 extern int memcmp (const void *__s1, const void *__s2, size_t __n)
            ^~~~~~
In file included from /usr/include/gtest/gtest.h:1874:0,
                 from /build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp:7:
/build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp: In member function ‘virtual void MSGPACKC_simple_buffer_str_fix_l_Test::TestBody()’:
/build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp:779:24: error: argument 1 null where non-null expected [-Werror=nonnull]
     EXPECT_EQ(0, memcmp(str, obj.via.str.ptr, str_size));
                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /build/msgpack-c/src/msgpack-2.1.2/include/msgpack/pack_define.h:15:0,
                 from /build/msgpack-c/src/msgpack-2.1.2/include/msgpack/pack.h:13,
                 from /build/msgpack-c/src/msgpack-2.1.2/include/msgpack.h:19,
                 from /build/msgpack-c/src/msgpack-2.1.2/test/msgpack_c.cpp:1:
/usr/include/string.h:66:12: note: in a call to function ‘int memcmp(const void*, const void*, size_t)’ declared here
 extern int memcmp (const void *__s1, const void *__s2, size_t __n)
            ^~~~~~
```

Since you cannot reproduce with the same GCC version, it really looks like it's related to the gtest version.

**Comment by redboltz** [comment]
@zorun ,thank you for the comment.

I just merged #602. Current master includes other fixes about gcc7 warnings. Some of them are depended on environment. 
Could you try current master branch on your environment with -Werror?

**Comment by zorun** [comment]
Yes, it works fine with the current master, thanks!
